compactcert/signer.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2019-2020 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package compactcert
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/ledger"
+)
+
+// sigFromAddr encapsulates a signature on a block header, which
+// will eventually be used to form a compact certificate for that
+// block.
+type sigFromAddr struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Signer basics.Address          `codec:"signer"`
+	Round  basics.Round            `codec:"rnd"`
+	Sig    crypto.OneTimeSignature `codec:"sig"`
+}
+
+func (ccw *Worker) signer() {
+	var sigkeys []crypto.OneTimeSignatureVerifier
+	var nextrnd basics.Round
+
+	err := ccw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		sigkeys, nextrnd, err = getSignedLast(tx)
+		return
+	})
+
+	if err != nil {
+		nextrnd := ccw.ledger.Latest() + 1
+		ccw.log.Warnf("ccw.signer(): using nextrnd=%d, cannot obtain last signed: %v", nextrnd, err)
+	} else {
+		// Check if we have any keys in common with the last signed set.
+		// If so, try to sign the next round after that.  Otherwise, if
```

Comment:
I expect it should be `<= ccw.ledger.Latest()`

Reply:
I am not able to follow this code.

Here is what I understand: please correct me.
nextrnd is determining where to start signing.
If getSignedLast returns an error, or if there is no overlap between last signed keys and the worker account keys, then nextrnd = ccw.ledger.Latest() + 1, otherwise, it will  start earlier. 

So, if there is an error (getSignedLast), then we start signing from a later round? Why is this?

Reply:
Sorry for the confusion here!  I tried to add some description to the schema (in `db.go`) which might help make this less confusing.

Your understanding is correct.  The role of the stored `signed_last` information is to make sure that, when this node starts running, it will resume signing blocks where it last stopped.  In particular, if this node was stopped, and in the meantime many blocks appeared, on restart this node will sign those new blocks.

If there's an error, we don't know where we last left off signing, so we won't sign blocks in this missing window.  That doesn't seem too bad, since compact certificates can be formed even if not every node signs.  Since we don't know where we last left off, the only real other option is to block this node's startup altogether, which seems less desirable.